rhp/v4/rpc.go

```diff
@@ -0,0 +1,705 @@
+	s := t.DialStream(ctx)
+	defer s.Close()
+
+	if err := rhp4.WriteRequest(s, rpcID, req); err != nil {
+		return fmt.Errorf("failed to write request: %w", err)
+	} else if err := rhp4.ReadResponse(s, resp); err != nil {
+		return fmt.Errorf("failed to read response: %w", err)
+	}
+	return nil
+}
+
+// RPCSettings returns the current settings of the host.
+func RPCSettings(ctx context.Context, t TransportClient) (rhp4.HostSettings, error) {
+	var resp rhp4.RPCSettingsResponse
+	err := callSingleRoundtripRPC(ctx, t, rhp4.RPCSettingsID, nil, &resp)
+	return resp.Settings, err
+}
+
+// RPCReadSector reads a sector from the host.
+func RPCReadSector(ctx context.Context, t TransportClient, prices rhp4.HostPrices, token rhp4.AccountToken, w io.Writer, root types.Hash256, offset, length uint64) (RPCReadSectorResult, error) {
+	req := &rhp4.RPCReadSectorRequest{
+		Prices: prices,
+		Token:  token,
+		Root:   root,
+		Offset: offset,
+		Length: length,
+	}
+	if err := req.Validate(t.PeerKey()); err != nil {
+		return RPCReadSectorResult{}, fmt.Errorf("invalid request: %w", err)
+	}
+
+	s := t.DialStream(ctx)
+	defer s.Close()
+
+	if err := rhp4.WriteRequest(s, rhp4.RPCReadSectorID, req); err != nil {
+		return RPCReadSectorResult{}, fmt.Errorf("failed to write request: %w", err)
+	}
+
+	var resp rhp4.RPCReadSectorStreamedResponse
+	if err := rhp4.ReadResponse(s, &resp); err != nil {
+		return RPCReadSectorResult{}, fmt.Errorf("failed to read response: %w", err)
+	}
+
+	// TODO: verify proof
+	n, err := io.Copy(w, io.LimitReader(s, int64(resp.DataLength)))
+	if err != nil {
+		return RPCReadSectorResult{}, fmt.Errorf("failed to read data: %w", err)
+	} else if n != int64(resp.DataLength) {
+		return RPCReadSectorResult{}, io.ErrUnexpectedEOF
+	}
+	return RPCReadSectorResult{
+		Cost: prices.RPCReadSectorCost(length),
+	}, nil
+}
+
+// RPCWriteSector writes a sector to the host.
+func RPCWriteSector(ctx context.Context, t TransportClient, prices rhp4.HostPrices, token rhp4.AccountToken, rl ReaderLen, duration uint64) (RPCWriteSectorResult, error) {
+	length, err := rl.Len()
+	if err != nil {
+		return RPCWriteSectorResult{}, fmt.Errorf("failed to get length: %w", err)
+	}
+
+	req := rhp4.RPCWriteSectorStreamingRequest{
+		Prices:     prices,
+		Token:      token,
+		Duration:   duration,
+		DataLength: uint64(length),
+	}
+
+	if err := req.Validate(t.PeerKey(), req.Duration); err != nil {
+		return RPCWriteSectorResult{}, fmt.Errorf("invalid request: %w", err)
+	}
+
+	stream := t.DialStream(ctx)
+	defer stream.Close()
+
+	bw := bufio.NewWriterSize(stream, t.FrameSize())
+
+	if err := rhp4.WriteRequest(bw, rhp4.RPCWriteSectorID, &req); err != nil {
+		return RPCWriteSectorResult{}, fmt.Errorf("failed to write request: %w", err)
+	}
+
+	sr := io.LimitReader(rl, int64(req.DataLength))
+	tr := io.TeeReader(sr, bw)
+	if req.DataLength < rhp4.SectorSize {
```

Comment:
Should we treat the case where `DataLength` is 0 and we're writing an empty sector?